lib/mapping/MapFormatH3M.cpp

```diff
@@ -220,15 +220,24 @@ void CMapLoaderH3M::readPlayerInfo()
 		else
 			totalFactions--; //exclude conflux for ROE
 
-		for(int fact = 0; fact < totalFactions; ++fact)
+		const bool isFactionRandom = mapHeader->players[i].isFactionRandom = reader.readBool();
+		const ui16 allFactionsMask = (mapHeader->version == EMapFormat::ROE)
+			? 0b1111111
+			: 0b11111111;
+		const bool allFactionsAllowed = mapHeader->version == EMapFormat::VCMI
+			|| (isFactionRandom && ((allowedFactions & allFactionsMask) == allFactionsMask));
+
+		if(!allFactionsAllowed)

```

Comment:
Interesting why the previous code did not work. Logically I see two cases
- all factions allowed
- some factions allowed
For each case we can have isFactionRandom set or unset. Are you sure we cover all these? Four cases in total.
For instance randomFaction + not all factions allowed?

Reply:
Are you sure we will not get conflux town for ROE when all factions allowed?

Reply:
1) Because `mapHeader->players[i].allowedFactions` by default contains all factions include mods. Then they tried to remove only vanila factions (which are not allowed). Thus, custom factions remained here in any way.

2) Then we should determine if all factions are allowed or not. 

For Non-VCMI maps:
- If a faction is not random let's suppose mod factions are not allowed even bit mask is full.
- If a faction is random all factions (include mods) are allowed when we have full bit mask

3) HoTa allows `Conflux` and `Cove` for RoE maps when a faction is random and all RoE factions are allowed
![roe_random](https://user-images.githubusercontent.com/58867270/140786930-af3361d5-d72d-453c-8e92-f200139c2d0a.jpg)
`All For One` - one of RoE maps. So, let's use HoTa implementation as an example.